### PR TITLE
mount rpool/nixos/var too

### PR DESCRIPTION
**Pull request by motiejus**
It is quite confusing to have a dataset `rpool/nixos/var`, which is not normally mounted.

Also see https://github.com/ne9z/dotfiles-flake/pull/6

### Discussion

I see two options here:

1. mark it `mountpoint=none` instead of `legacy` to clearly show that it's not meant to be mounted.
2. mount the dataset, as this PR does.

Let me know what your prefer.

### CODE REVIEW

**Comment by ne9z** [comment]
Thanks for the suggestion!  However, I think it is not really necessary to
mount /var.  Let me explain my motivation of this.

In writing the guide, I strive to not complicate or fragment the ZFS
dataset layout or the mountpoints.  The existence of each and every
additional dataset and its children must be justified by some
necessity.  With this approach, we would start out with one single
dataset
```
  rpool -> /
```
and see if that is sufficient for our purpose, which is to install
NixOS on it.  As it turns out, the GRUB bootloader only supports a
fixed set of ZFS pool features.  Not to be limited by this constraint
on the main system, we create an additional pool for GRUB and /boot::
```
  rpool -> /
  bpool -> /boot
```
Note that the name ``rpool`` refers to two different concepts, one is
the pool ``rpool``, the other is the dataset ``rpool``. It turned out
that the dataset ``rpool`` has a few limitations.  First, it can not
be renamed by "zfs rename"; it can only be renamed by a reimport of
the pool.  Furthermore, it can not be reencrypted in the event of a
breach of the encryption password.  Therefore, we use ``rpool`` and
``bpool`` as placeholders, or, containers, as I've came to call them.
They are not mounted, and ``canmount=off`` property is set.  This
gives us::
```
  rpool/nixos -> /
  bpool/nixos -> /boot
```
Here, my experience with other distros and the concept of alternative
roots, called "boot environment", made me to add one more child
dataset::
```
  rpool/nixos/root -> /
  bpool/nixos/root -> /boot
```
The intention is to use them in this way::
```
  rpool/debian/11.7 -> /
  bpool/debian/11.7 -> /boot
  rpool/debian/12.1 -> /
  bpool/debian/12.1 -> /boot
```
but, Nix is so nice that it supports rollback at the OS level, we do
not use it for NixOS.  The last ``nixos/root`` component is therefore
a vestige of me writing guides for other distros, namely alpine, arch
and fedora.

In the last step, we want to make the layout compatible with a
immutable root filesystem.  There are many ways to accomplish this;
I've chosen the method of showing mountpoints in the dataset
hierarchy.  For example, we need to keep the contents of ``/home``, so
we add a separate dataset for it and arrange it to be mounted in the
system configuration::
```
  rpool/nixos/home -> /home
  rpool/nixos/root -> /
  bpool/nixos/root -> /boot
```
For logs and data generated by various programs, such as VM image from
libvirt, we also need to prevent ``/var/log`` and ``/var/lib`` from
being wiped on reboot.  I could use something like this::
```
  rpool/nixos/myLog -> /var/log
  rpool/nixos/programData -> /var/lib
```
but, in the same spirit of mounting ``/home``  at ``nixos/home``, I
used::
```
  rpool/nixos/var/log -> /var/log
  rpool/nixos/var/lib -> /var/lib
```
As other directories under ``/var`` are mainly temporary and therefore
should be wiped on reboot, I've made it a container, i.e., a
placeholder in dataset layout and not mounted.  This is also to
prevent fragmenting the root ``/`` too much.

If really needed, we can create separate datasets for each of the system
directories::
```
  bin  boot  dev  etc  home  nix  root  run  srv  tmp  usr  var
```
But I think it might be too much and unnecessary for a new user.

In short, do you think other sub-directories of ``/var`` is worth preserving
between reboots?  I and maintainers of Ubuntu[1] don't think so,
therefore I made the choice of only mounting ``/var/log`` and ``/var/lib``.

[1]: See [here](https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2022.04%20Root%20on%20ZFS.html), in particular they created ``rpool/ROOT/ubuntu_$UUID/var`` with ``canmount=off`` and suggested optional child datasets for NetworkManager, Docker, etc.

**Comment by motiejus** [comment]
> <...>
> As other directories under `/var` are mainly temporary and therefore should be wiped on reboot, I've made it a container, i.e., a placeholder in dataset layout and not mounted. This is also to prevent fragmenting the root `/` too much.

This all makes total sense, thanks for a wide explanation!

> In short, do you think other sub-directories of `/var` is worth preserving between reboots? I and maintainers of Ubuntu[1] don't think so, therefore I made the choice of only mounting `/var/log` and `/var/lib`.

To be frank, I did not consider immutable root, but I completely agree with the statement above (yours and Ubuntu maintainers).

> [1]: See [here](https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2022.04%20Root%20on%20ZFS.html), in particular they created `rpool/ROOT/ubuntu_$UUID/var` with `canmount=off` and suggested optional child datasets for NetworkManager, Docker, etc.

Can we do this for our guide then (that's the option 2 of my suggestion in the PR):

```patch
-zfs create -o mountpoint=legacy  rpool/nixos/var
+zfs create -o mountpoint=none  rpool/nixos/var
```

That way, it's clear to the user that `/var` is not meant to be mounted. If you agree, I will update the PR.

**Comment by ne9z** [comment]
> Can we do this for our guide then (that's the option 2 of my suggestion in the PR):
> 
> -zfs create -o mountpoint=legacy  rpool/nixos/var
> +zfs create -o mountpoint=none  rpool/nixos/var
> 
> That way, it's clear to the user that /var is not meant to be mounted. If you agree, I will update the PR.

Sure!  It is definitely an improvement.

**Review by ne9z — APPROVED** [review]
